library/src/main/java/bio/terra/pfb/PfbReader.java

```diff
@@ -0,0 +1,211 @@
+      }
+      return data;
+    } catch (IOException e) {
+      return List.of(getErrorMessage(e));
+    }
+  }
+
+  public Metadata getPFBMetadata(String fileLocation) throws IOException {
+    File pfbData = new File(fileLocation);
+    // Deserialize the above generated avro data file
+    DatumReader<Entity> datumReader = new SpecificDatumReader<>(Entity.class);
+    Entity data = null;
+    Metadata result = null;
+    try (DataFileReader<Entity> dataFileReader = new DataFileReader<>(pfbData, datumReader)) {
+      // A PFB Avro file consists of a list of "Entity" objects (Defined in sample.advl)
+      // One of these entities must be a "Metadata" object (Also defined in sample.advl)
+      // The rest of the entities are the data entries
+      // Assuming the first object is the metadata object
+      data = dataFileReader.next(data);
+      result = (Metadata) data.getObject();
+      if (result != null) {
+        return result;
+      }
+    }
+    throw new InvalidPfbException("Error reading PFB Metadata object");
+  }
+
+  Schema readUrlPFBSchema(String signedUrl) throws IOException {
+    Schema schema;
+    DatumReader<Entity> datumReader = new SpecificDatumReader<>(Entity.class);
+    try (InputStream in = readFromSignedUrl(signedUrl);
+        DataFileStream<Entity> reader = new DataFileStream<>(in, datumReader)) {
+      schema = reader.getSchema();
+    } catch (IOException ex) {
+      throw new IOException("Error reading PFB file from signed URL", ex);
+    }
+    return schema;
+  }
+
+  // Helper methods
+
+  private String getErrorMessage(Exception e) {
+    return "Error: " + e.getMessage();
+  }
+
+  boolean isValidUrl(String fileLocation) {
+    try {
+      new URL(fileLocation);
+      return true;
+    } catch (IOException e) {
+      return false;
+    }
+  }
+
+  InputStream readFromSignedUrl(String signedUrl) throws IOException {
+    URL urlObject = new URL(signedUrl);
+    URLConnection urlConnection = urlObject.openConnection();
+    return urlConnection.getInputStream();
+  }
+
+  /* This method will be removed in AJ-1288
+       Need to instead encode/decode enum typed variables
+     From docs (https://github.com/uc-cdis/pypfb/blob/master/docs/index.md#enum):
+    "Because Avro can't store anything except alphanumeric and _ symbols
+    (https://avro.apache.org/docs/1.9.1/spec.html#Enums), all enums are encoded in
+    such a way, that all other symbols is being encoded with codepoint wrapped in
+    single underscores. For example bpm > 60 will be encoded in: bpm_32__62__32_60,
+    so space   is encoded as _32_ and greater sign > into _62_. Same for Unicode
+    characters: ä - _228_, ü - _252_. The Avro schema also doesn't allow for the
+    first character to be a number. So we encode the first character in the way if
+    the character happens to be a number."
+  */
+  private String convertEnum(String schema) {
+    HashMap<String, String> map = new HashMap<>();
```

Comment:
Is there a reference / source of truth somewhere that we can check going forward to make sure this map is comprehensive?  I imagine it could become outdated if additional characters/encodings were introduced.

Reply:
Definitely not the best solution! This code is a holdover until addressed the root issue is addressed in AJ-1288.